File: Data Structures & Algorithms/combination-target-sum/submission-5.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        ret = []
        currComb = []


        def dfs(i, remainder):
            if i == len(nums):
                if remainder == 0: ret.append(currComb[:])
            else:
                mult = 1

                while remainder - mult * nums[i] >= 0: 
                    currComb.append(nums[i])
                    dfs(i+1, remainder - mult * nums[i])
                    mult += 1
                
                while currComb and currComb[-1] == nums[i]: currComb.pop()

                dfs(i+1, remainder)
        
        dfs(0, target)
        return ret

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-5.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int L = nums.length;

        int[] ret = new int[L];
        int[] prefix = new int[L];
        int[] suffix = new int[L];

        prefix[0] = 1;
        suffix[L-1] = 1;

        for (int i = 1; i < L; i++) {
            prefix[i] = prefix[i-1] * nums[i-1];
        }

        for (int i = L - 2; i >= 0; i--) {
            suffix[i] = suffix[i+1] * nums[i+1];
        }

        for (int i = 0; i < L; i++) {
            ret[i] = prefix[i] * suffix[i];
        }

        return ret;
    }
}  
